jina/executors/indexers/vector.py

```diff
@@ -69,10 +70,73 @@ def __init__(self,
                                 f'compress_level overriden from `ref_indexer` to {ref_indexer.compress_level} \n'
                                 f'index_filename overriden from `ref_indexer` to {ref_indexer.index_filename}')
             self.ref_indexer_workspace_name = ref_indexer.workspace_name
+            # if ref indexer didn't have it assume default False
+            self.delete_on_save = ref_indexer.__dict__.get('delete_on_save') or delete_on_save
+
+    def _delete_invalid_indices(self):
+        valid = self.valid_indices[self.valid_indices == True]  # noqa
+        if len(valid) != len(self.valid_indices):
+            self._clean_memmap()
+            self._post_clean_memmap(valid)
+
+    def _post_clean_memmap(self, valid):
+        # here we need to make sure the fields
+        # that depend on the valid_indices are cleaned up too
+        valid_key_bytes = np.frombuffer(self.key_bytes, dtype=(np.str_, self.key_length))[
+            self.valid_indices].tobytes()
+        self.key_bytes = valid_key_bytes
+        self._size = len(valid)
+        self.valid_indices = valid
+        del self.__dict__['CACHED__int2ext_id']
+        del self.__dict__['CACHED__ext2int_id']
+
+    def _clean_memmap(self):
+        # clean up the underlying matrix of entries marked for deletion
+        # first we need to make sure we flush the writing handler
+        if self.write_handler and not self.write_handler.closed:
+            print(f'we flush and close')
+            self.write_handler.flush()
+            self.write_handler.close()
+            self.handler_mutex = False
+        # we need an intermediary file
+        tmp_path = self.index_abspath + 'tmp'
+        # f = open(tmp_path, 'wb')
+        # f.close()
+        # force the raw_ndarray (underlying matrix) to re-read from disk
+        # (needed when there were writing ops to be flushed)
+        del self.__dict__['CACHED__raw_ndarray']
+        filtered = self._raw_ndarray[self.valid_indices]
+
+        # write the bytes in the respective files
+        if self.compress_level > 0:
+            new_gzip_fh = gzip.open(tmp_path, 'wb', compresslevel=self.compress_level)
+            new_gzip_fh.write(filtered.tobytes())
+            new_gzip_fh.close()
+            del new_gzip_fh
+        else:
+            filtered_data_fh = open(tmp_path, 'wb')
```

Comment:
use context manager here for safe handling

Reply:
Will do a clean-up